constraints/miden-vm/decoder.air

```diff
@@ -0,0 +1,483 @@
+
+    # Enforce that when SPLIT or LOOP operation is executed, the top of the operand stack must 
+    # contain a binary value.
+    enf is_binary([s0]) when f_split(op_bits) | f_loop(op_bits)
+
+    # Enforce that When REPEAT operation is executed, the value at the top of the operand stack 
+    # must be 1.
+    enf s0 = 1 when f_repeat(op_bits, extra)
+
+    # Enforce that when REPEAT operation is executed, the value in hasher[4] column (the 
+    # is_loop_body flag), must be set to 1. This ensures that REPEAT operation can be executed only
+    # inside a loop.
+    enf hasher[4] = 1 when f_repeat(op_bits, extra)
+
+    # Enforce that when RESPAN operation is executed, we need to make sure that the block ID is 
+    # incremented by 8.
+    enf addr' = addr + 8 when f_respan(op_bits, extra)
+
+    # Enforce that when END operation is executed and we are exiting a loop block (i.e., is_loop, 
+    # value which is stored in hasher[5], is 1), the value at the top of the operand stack must be 
+    # 0.
+    enf s0 = 0 when f_end(op_bits, extra) & hasher[5]
+
+    # Enforce that when END operation is executed and the next operation is REPEAT, values in 
+    # hasher[0], ..., hasher[4] (the hash of the current block and the is_loop_body flag) must be 
+    # copied to the next row.
+    enf is_unchanged([hasher[i]]) when f_end(op_bits, extra) & f_repeat_next(op_bits, extra) for i in 0..5
+
+    # Enforce that a HALT instruction can be followed only by another HALT instruction.
+    enf f_halt(op_bits, extra) * binary_not(f_halt_next) = 0
+
+    # Enforce that when a HALT operation is executed, block address column (addr) must be 0.
+    enf addr = 0 when f_halt(op_bits, extra)
+
+    # Enforce that values in op_bits columns must be binary.
+    enf is_binary([b]) for b in op_bits
+
+    # Enforce that when the value in is_span column is set to 1, control flow operations cannot be 
+    # executed on the VM, but when is_span flag is 0, only control flow operations can be executed
+    # on the VM.
+    enf 1 - is_span - f_ctrl = 0
+
+
+# Enforces constraint for computing block hashes.
+ev block_hash_computation_constraints(main: [addr, op_bits[7], hasher[8]], aux: [extra, p[4]]):
+    # Operation label for HASHER_LINER_HASH operation.
+    let op_linhash = 3
+
+    # Operation label for HASHER_RETURN_HASH operation.
+    let op_hout = 1
+
+    # Label specifying that we are starting a new hash computation.
+    let m_bp = get_transition_label(op_linhash)
+
+    # Label specifying that we are absorbing the next sequence of 8 elements into an ongoing hash 
+    # computation.
+    let m_abp = get_transition_label(op_linhash)
+
+    # Label specifying that we are reading the result of a hash computation.
+    let m_hout = get_transition_label(op_hout)
+
+    # `alpha` is inherited random values array.
+    let absorb_sum = sum([alpha[i + 8] * hasher[i] for i in 0..8])
+    let return_sum = sum([alpha[i + 8] * hasher[i] for i in 0..4])
```

Comment:
Let's change the names here to match with the nomenclature for the state of the hash function.

- "capacity" portion of the state: the first 4 elements
- "rate" portion of the state: elements 5-12
- "digest" portion of the state: the output, elements 5-8

So I would rename `absorb_sum` as `rate_sum` and `return_sum` as `digest_sum`

This is consistent with the naming we use [here](https://github.com/0xPolygonMiden/miden-vm/blob/159a04f3d4648f77d664642409e521438378bd6c/core/src/chiplets/hasher.rs#L23)